### ISSUE TITLE
DBCA loses contact during first run

### ISSUE THREAD

**Issue by neerolyte** [oracle/docker-images#270]
I'm not sure why this has started happening, I upgraded Docker (to 1.13.0 on OS X) and now when I run an Oracle container it fails pretty early during DBCA:

```
$ docker run -it --name oracle-db oracle/database:12.1.0.2-ee
ORACLE AUTO GENERATED PASSWORD FOR SYS, SYSTEM AND PDBAMIN: kjFwWpzHVNA=1

LSNRCTL for Linux: Version 12.1.0.2.0 - Production on 08-FEB-2017 20:00:54

Copyright (c) 1991, 2014, Oracle.  All rights reserved.

Starting /opt/oracle/product/12.1.0.2/dbhome_1/bin/tnslsnr: please wait...

TNSLSNR for Linux: Version 12.1.0.2.0 - Production
System parameter file is /opt/oracle/product/12.1.0.2/dbhome_1/network/admin/listener.ora
Log messages written to /opt/oracle/diag/tnslsnr/6d4cfaec1cf6/listener/alert/log.xml
Listening on: (DESCRIPTION=(ADDRESS=(PROTOCOL=ipc)(KEY=EXTPROC1)))
Listening on: (DESCRIPTION=(ADDRESS=(PROTOCOL=tcp)(HOST=0.0.0.0)(PORT=1521)))

Connecting to (DESCRIPTION=(ADDRESS=(PROTOCOL=IPC)(KEY=EXTPROC1)))
STATUS of the LISTENER
------------------------
Alias                     LISTENER
Version                   TNSLSNR for Linux: Version 12.1.0.2.0 - Production
Start Date                08-FEB-2017 20:00:55
Uptime                    0 days 0 hr. 0 min. 0 sec
Trace Level               off
Security                  ON: Local OS Authentication
SNMP                      OFF
Listener Parameter File   /opt/oracle/product/12.1.0.2/dbhome_1/network/admin/listener.ora
Listener Log File         /opt/oracle/diag/tnslsnr/6d4cfaec1cf6/listener/alert/log.xml
Listening Endpoints Summary...
  (DESCRIPTION=(ADDRESS=(PROTOCOL=ipc)(KEY=EXTPROC1)))
  (DESCRIPTION=(ADDRESS=(PROTOCOL=tcp)(HOST=0.0.0.0)(PORT=1521)))
The listener supports no services
The command completed successfully
Copying database files
1% complete
3% complete
DBCA Operation failed.
Look at the log file "/opt/oracle/cfgtoollogs/dbca/ORCLCDB/ORCLCDB.log" for further details.
Database Configuration Assistant failed to detect options for this Oracle Installation.
ORA-12547: TNS:lost contact

Due to this failure some Database Configuration Assistant features will not be available. Do you want to continue?

Unique database identifier check passed.

/opt/oracle/oradata/ has enough space. Required space is 7665 MB , available space is 46624 MB.
/ has enough space. Required space is 0 MB , available space is 46624 MB.
File Validations Successful.
Copying database files
DBCA_PROGRESS : 1%
ORA-12547: TNS:lost contact

DBCA_PROGRESS : 3%
Error while cataloging RMAN Backups
DBCA_PROGRESS : DBCA Operation failed.

SQL*Plus: Release 12.1.0.2.0 Production on Wed Feb 8 20:01:46 2017

Copyright (c) 1982, 2014, Oracle.  All rights reserved.

ERROR:
ORA-12547: TNS:lost contact


Enter user-name: SP2-0306: Invalid option.
Usage: CONN[ECT] [{logon|/|proxy} [AS {SYSDBA|SYSOPER|SYSASM|SYSBACKUP|SYSDG|SYSKM}] [edition=value]]
where <logon> ::= <username>[/<password>][@<connect_identifier>]
      <proxy> ::= <proxyuser>[<username>][/<password>][@<connect_identifier>]
Enter user-name: SP2-0306: Invalid option.
Usage: CONN[ECT] [{logon|/|proxy} [AS {SYSDBA|SYSOPER|SYSASM|SYSBACKUP|SYSDG|SYSKM}] [edition=value]]
where <logon> ::= <username>[/<password>][@<connect_identifier>]
      <proxy> ::= <proxyuser>[<username>][/<password>][@<connect_identifier>]
SP2-0157: unable to CONNECT to ORACLE after 3 attempts, exiting SQL*Plus
mv: cannot stat '/opt/oracle/product/12.1.0.2/dbhome_1/dbs/spfileORCLCDB.ora': No such file or directory
mv: cannot stat '/opt/oracle/product/12.1.0.2/dbhome_1/dbs/orapwORCLCDB': No such file or directory
#########################
DATABASE IS READY TO USE!
#########################
tail: cannot open '/opt/oracle/diag/rdbms/*/*/trace/alert*.log' for reading: No such file or directory
tail: no files remaining
```

I haven't modified the docker-images checkout in any way, I've tried rebuilding from the 4f25e0dbac2bb982dda2e778f15c35866afd0245 (the current master commit) and get the same.

I've put the build log in a gist here: https://gist.github.com/neerolyte/89fe09a713e0965763d67a4ae5b008ad

The only thing jumping out is the segfault:

```
/opt/oracle/install/installDBBinaries.sh: line 63:  2860 Segmentation fault      $ORACLE_HOME/perl/bin/perl -v
```

but looking through commit history - that looks to be intentional (from https://github.com/oracle/docker-images/issues/123).

I've put these zip files in the dockerfiles dir before building:

```
$ md5sum OracleDatabase/dockerfiles/12.1.0.2/linuxamd64_12102_database_?of2.zip
080435a40bd4c8dff6399b231a808e9a  OracleDatabase/dockerfiles/12.1.0.2/linuxamd64_12102_database_1of2.zip
30f20ef9437442b8282ce3984546c982  OracleDatabase/dockerfiles/12.1.0.2/linuxamd64_12102_database_2of2.zip
```

**Comment by gvenzl** [comment]
Hey @neerolyte, looks like you are running in a nasty little Perl problem with Intel HD Graphics. Hence why we check whether Perl is executable and if not, recompile the latest into the image. Looks like that up to the download this all worked but then there were some issues during the complication. Not sure whether that's the issue. Can you try and execute `/opt/oracle/product/12.1.0.2/dbhome_1/perl/bin/perl -v` in your image and see whether it works as expected?
If it does, we will need to have a look at `/opt/oracle/cfgtoollogs/dbca/ORCLCDB/ORCLCDB.log` from your failing container and see what's going on.

**Comment by neerolyte** [comment]
Perl seems ok:

```
[oracle@1fdb49ead3b7 ~]$ /opt/oracle/product/12.1.0.2/dbhome_1/perl/bin/perl -v

This is perl 5, version 14, subversion 1 (v5.14.1) built for x86_64-linux-thread-multi
[...]
[oracle@1fdb49ead3b7 ~]$ echo $?
0
```

`/opt/oracle/cfgtoollogs/dbca/ORCLCDB/ORCLCDB.log` didn't seem to have anything important, so I hadn't included it in the first issue:

```
[oracle@f13b735353fd ~]$ cat /opt/oracle/cfgtoollogs/dbca/ORCLCDB/ORCLCDB.log
Database Configuration Assistant failed to detect options for this Oracle Installation.
ORA-12547: TNS:lost contact

Due to this failure some Database Configuration Assistant features will not be available. Do you want to continue?

Unique database identifier check passed.

/opt/oracle/oradata/ has enough space. Required space is 7665 MB , available space is 37378 MB.
/ has enough space. Required space is 0 MB , available space is 37378 MB.
File Validations Successful.
Copying database files
DBCA_PROGRESS : 1%
ORA-12547: TNS:lost contact

DBCA_PROGRESS : 3%
Error while cataloging RMAN Backups
DBCA_PROGRESS : DBCA Operation failed.
```

p.s I am still poking other bits of the repo and if I figure anything out I'll add a comment.

**Comment by neerolyte** [comment]
There's some odd broken symlinks, I'm not sure if it's a cause or effect thing at this point though:

```
[oracle@de9e89aa7f95 ~]$ bash -x /opt/oracle/runOracle.sh
[...]
File Validations Successful.
Copying database files
DBCA_PROGRESS : 1%
ORA-12547: TNS:lost contact

DBCA_PROGRESS : 3%
Error while cataloging RMAN Backups
DBCA_PROGRESS : DBCA Operation failed.
[...]
+ moveFiles
+ '[' '!' -d /opt/oracle/oradata/dbconfig/ORCLCDB ']'
+ mkdir -p /opt/oracle/oradata/dbconfig/ORCLCDB/
+ mv /opt/oracle/product/12.1.0.2/dbhome_1/dbs/spfileORCLCDB.ora /opt/oracle/oradata/dbconfig/ORCLCDB/
mv: cannot stat '/opt/oracle/product/12.1.0.2/dbhome_1/dbs/spfileORCLCDB.ora': No such file or directory
+ mv /opt/oracle/product/12.1.0.2/dbhome_1/dbs/orapwORCLCDB /opt/oracle/oradata/dbconfig/ORCLCDB/
mv: cannot stat '/opt/oracle/product/12.1.0.2/dbhome_1/dbs/orapwORCLCDB': No such file or directory
+ mv /opt/oracle/product/12.1.0.2/dbhome_1/network/admin/tnsnames.ora /opt/oracle/oradata/dbconfig/ORCLCDB/
[...]
[oracle@de9e89aa7f95 ~]$ ls -la /opt/oracle/product/12.1.0.2/dbhome_1/dbs/spfileORCLCDB.ora
lrwxrwxrwx 1 oracle dba 54 Feb 11 07:19 /opt/oracle/product/12.1.0.2/dbhome_1/dbs/spfileORCLCDB.ora -> /opt/oracle/oradata/dbconfig/ORCLCDB/spfileORCLCDB.ora
[oracle@de9e89aa7f95 ~]$ ls -la /opt/oracle/oradata/dbconfig/ORCLCDB/spfileORCLCDB.ora
ls: cannot access /opt/oracle/oradata/dbconfig/ORCLCDB/spfileORCLCDB.ora: No such file or directory
```

**Comment by neerolyte** [comment]
So it's definitely the [dbca call](https://github.com/oracle/docker-images/blob/4abc5ae/OracleDatabase/dockerfiles/12.1.0.2/createDB.sh#L56) that's the first thing to fall over (I modified a [bunch of stuff](https://github.com/oracle/docker-images/compare/master...neerolyte:daves-debug-stuff) to an error handling mechanic I'm more comfortable with to track it down).

It seems to generate [this trace.log](https://gist.github.com/neerolyte/500f4187dd344d9549e0b489ed99b6e5) that contains more of the `ORA-12547: TNS:lost contact` errors - but I'm not clear if there's something else in there that explains what's failed. It looks like maybe the listener is dropping the connection when rman tries to do something (install some template DB?)?

I also tried stracing the dbca call, but didn't get anything out of that.

**Comment by neerolyte** [comment]
So I found out I could get the listener to generate a trace file with this modification to `createDB.sh`:

```bash
...
lsnrctl start
lsnrctl << EOF
set trc_level ADMIN
exit
EOF
if ! dbca -silent -responseFile $ORACLE_BASE/dbca.rsp; then
...
```

Which generates the following:

```
[root@a3bd971d33ee oracle]# cat /opt/oracle/diag/tnslsnr/a3bd971d33ee/listener/trace/ora_1088_140698670896640.trc
Trace file /opt/oracle/diag/tnslsnr/a3bd971d33ee/listener/trace/ora_1088_140698670896640.trc
2017-02-11 10:01:01.771318 : nsglrespond:entry
2017-02-11 10:01:01.771500 : nsdo:cid=5, opcode=67, *bl=82, *what=1, uflgs=0x0, cflgs=0x3
2017-02-11 10:01:01.771512 : nsdo:rank=64, nsctxrnk=0
2017-02-11 10:01:01.771520 : nsdo:nsctx: state=8, flg=0x2100400c, mvd=0
2017-02-11 10:01:01.771528 : nsdo:gtn=229, gtc=229, ptn=10, ptc=8111
2017-02-11 10:01:01.771536 : nsdo:82 bytes to NS buffer
2017-02-11 10:01:01.771543 : nsdo:nsctxrnk=0
2017-02-11 10:01:01.771552 : nsglrespond:exit
2017-02-11 10:01:01.771560 : nsgldissolve:entry
2017-02-11 10:01:01.771570 : nsvntx_dei:entry
2017-02-11 10:01:01.771578 : nsvntx_dei:exit
2017-02-11 10:01:01.771586 : nstimarmed:no timer allocated
2017-02-11 10:01:01.771594 : nsfull_cls:cid=5, opcode=65, *bl=0, *what=0, uflgs=0x0, cflgs=0x0
2017-02-11 10:01:01.771602 : nsfull_cls:nsctx: state=8, flg=0x21004008, mvd=82
2017-02-11 10:01:01.771612 : nsevunreg:cid=5, sgt=0, rdm=0
2017-02-11 10:01:01.771623 : nsevunreg:3 registered connection(s)
2017-02-11 10:01:01.771631 : nsdo:cid=5, opcode=67, *bl=0, *what=1, uflgs=0x0, cflgs=0x1
2017-02-11 10:01:01.771638 : nsdo:nsctx: state=8, flg=0x1004008, mvd=82
2017-02-11 10:01:01.771646 : nsdo:gtn=229, gtc=229, ptn=92, ptc=8111
2017-02-11 10:01:01.771654 : nsdofls:DATA flags: 0x40
2017-02-11 10:01:01.771661 : nsdofls:sending NSPTDA packet
2017-02-11 10:01:01.771669 : nspsend:plen=92, type=6
2017-02-11 10:01:01.771677 : ntuswrte:entry
2017-02-11 10:01:01.771687 : ntuswrte:Wrote on 14 bytes 92
2017-02-11 10:01:01.771694 : ntuswrte:exit
2017-02-11 10:01:01.771702 : nspsend:92 bytes to transport
2017-02-11 10:01:01.771712 : nsclose:closing transport
2017-02-11 10:01:01.771720 : ntusdisc:entry
2017-02-11 10:01:01.771731 : ntusdisc:Closed socket 14
2017-02-11 10:01:01.771739 : ntusdisc:exit
2017-02-11 10:01:01.771747 : nsclose:global context check-out (from slot 5) complete
2017-02-11 10:01:01.771755 : nsvntx_dei:entry
2017-02-11 10:01:01.771762 : nsvntx_dei:exit
2017-02-11 10:01:01.771771 : nsmfr:2784 bytes at 0x9e8920
2017-02-11 10:01:01.771779 : nsmfr:1912 bytes at 0x9e81a0
2017-02-11 10:01:01.771787 : nsgldissolve:Deallocating cxd 0x9ea010.
2017-02-11 10:01:01.771795 : nsgldissolve:exit
2017-02-11 10:01:01.771890 : nsevwait:3 registered connection(s)
2017-02-11 10:01:01.771901 : nsevwait:0 pre-posted event(s)
2017-02-11 10:01:01.771909 : nsevwait:waiting for transport event (1 thru 4)...
[root@a3bd971d33ee oracle]#
```

... which looks ok?

The listener is still running after that.

**Comment by neerolyte** [comment]
Figured it out, `relink all` (part of the perl compile) was failing due to no swap space. The non-zero exit status wasn't being caught in the `installPerl.sh` script.

Weirdly enough, just as I figured it out I was prompted to upgrade Docker, it upgraded to 1.13.1, and now I have swap space again (I did try creating a swapfile and `swapon`ing it in 1.13.0 before upgrading and that worked too).

I'll submit a PR for the error checking in a sec.
